src/core/ext/filters/client_channel/lb_policy/grpclb/load_balancer_api.cc

```diff
@@ -75,152 +81,94 @@ grpc_grpclb_request* grpc_grpclb_load_report_request_create(
                 "grpc_lb_v1_LoadBalanceResponse has too long token. len=%zu",
                 token.size);
       }
-      cur->drop = grpc_lb_v1_Server_drop(servers[i]);
+      cur.drop = grpc_lb_v1_Server_drop(servers[i]);
     }
   }
-  return server_list;
+  return true;
 }
 
-void grpc_grpclb_destroy_serverlist(grpc_grpclb_serverlist* serverlist) {
-  if (serverlist == nullptr) {
-    return;
-  }
-  for (size_t i = 0; i < serverlist->num_servers; i++) {
-    gpr_free(serverlist->servers[i]);
-  }
-  gpr_free(serverlist->servers);
-  gpr_free(serverlist);
+grpc_millis grpc_grpclb_duration_to_millis(
+    const google_protobuf_Duration* duration_pb) {
+  return static_cast<grpc_millis>(
+      (google_protobuf_Duration_seconds(duration_pb) * GPR_MS_PER_SEC) +
+      (google_protobuf_Duration_nanos(duration_pb) / GPR_NS_PER_MS));
 }
 
-grpc_grpclb_serverlist* grpc_grpclb_serverlist_copy(
-    const grpc_grpclb_serverlist* server_list) {
-  grpc_grpclb_serverlist* copy = static_cast<grpc_grpclb_serverlist*>(
-      gpr_zalloc(sizeof(grpc_grpclb_serverlist)));
-  copy->num_servers = server_list->num_servers;
-  copy->servers = static_cast<grpc_grpclb_server**>(
-      gpr_malloc(sizeof(grpc_grpclb_server*) * server_list->num_servers));
-  for (size_t i = 0; i < server_list->num_servers; i++) {
-    copy->servers[i] = static_cast<grpc_grpclb_server*>(
-        gpr_malloc(sizeof(grpc_grpclb_server)));
-    memcpy(copy->servers[i], server_list->servers[i],
-           sizeof(grpc_grpclb_server));
-  }
-  return copy;
-}
+}  // namespace
 
-bool grpc_grpclb_serverlist_equals(const grpc_grpclb_serverlist* lhs,
-                                   const grpc_grpclb_serverlist* rhs) {
-  if (lhs == nullptr || rhs == nullptr) {
-    return false;
-  }
-  if (lhs->num_servers != rhs->num_servers) {
-    return false;
+bool GrpcLbResponseParse(const grpc_slice& encoded_grpc_grpclb_response,
+                         upb_arena* arena, GrpcLbResponse* result) {
+  grpc_lb_v1_LoadBalanceResponse* response =
+      grpc_lb_v1_LoadBalanceResponse_parse(
+          reinterpret_cast<const char*>(
+              GRPC_SLICE_START_PTR(encoded_grpc_grpclb_response)),
+          GRPC_SLICE_LENGTH(encoded_grpc_grpclb_response), arena);
+  // Handle serverlist responses.
+  if (ParseServerList(*response, &result->serverlist)) {
+    result->type = result->SERVERLIST;
+    return true;
   }
-  for (size_t i = 0; i < lhs->num_servers; i++) {
-    if (!grpc_grpclb_server_equals(lhs->servers[i], rhs->servers[i])) {
-      return false;
-    }
+  // Handle initial responses.
+  auto* initial_response =
+      grpc_lb_v1_LoadBalanceResponse_initial_response(response);
+  if (initial_response != nullptr) {
+    result->type = result->INITIAL;
+    const google_protobuf_Duration* client_stats_report_interval =
+        grpc_lb_v1_InitialLoadBalanceResponse_client_stats_report_interval(
+            initial_response);
+    result->client_stats_report_interval =
```

Comment:
question: do we need to keep the previous check that client_stats_report_interval parsed correctly?, i.e. that it's not nullptr?

Or is it safe to assume we're ok since the whole initial response already parsed?

Reply:
Good catch!  You're right, we should check that.  But I don't think it's a parsing issue; the check is for the case where the field is unset.

Done.